ocaml/xapi/helpers.ml

```diff
@@ -2119,3 +2119,219 @@ let get_active_uefi_certificates ~__context ~self =
       custom_uefi_certs
 
 let uefi_mode_to_string = function `setup -> "setup" | `user -> "user"
+
+module BoundedPsq = struct
+  module type Ordered = sig
+    type t
+
+    val compare : t -> t -> int
+  end
+
+  module type S = sig
+    type t
+
+    type k
+
+    type v
+
+    val create : capacity:int -> t
+
+    val add : t -> k -> v -> unit
+    (** [add t k v] adds mapping [k] => [v] to the priority search queue. If
+        the addition of this mapping would exceed the capacity of the queue,
+        the highest priority (lowest) entry is removed to make space
+        for the newest entry.
+
+        If an entry for [k] is already present, the extant entry is
+        updated with [v]. *)
+
+    val remove : t -> k -> unit
+
+    val clear : t -> unit
+
+    val min : t -> (k * v) option
+
+    val find_opt : k -> t -> v option
+
+    val contains : t -> k -> bool
+
+    val iter : (k -> v -> unit) -> t -> unit
+
+    val size : t -> int
+  end
+
+  module Make (K : Ordered) (V : Ordered) :
+    S with type k = K.t and type v = V.t = struct
+    module Q = Psq.Make (K) (V)
+
+    type k = Q.k
+
+    type v = Q.p
+
+    type t = {capacity: int; mutable queue: Q.t}
```

Comment:
How about to make the `BoundedPsq` a pure functional module and move the `mutability` to `AuthenticationCache`?